usaspending_api/etl/management/commands/elasticsearch_indexer.py

```diff
@@ -3,232 +3,211 @@
-            "--skip-counts",
-            action="store_true",
-            help="When this flag is set, the ETL process will skip the record counts to reduce operation time",
+            help="When this flag is set, the script will include the extra steps"
+            "to calculate deleted records and remove from the target index",
         )
         parser.add_argument(
             "--index-name",
             type=str,
             help="Provide name for new index about to be created. Only used when --create-new-index is provided",
+            metavar="",
         )
         parser.add_argument(
             "--create-new-index",
             action="store_true",
             help="It needs a new unique index name and set aliases used by API logic to the new index",
         )
-        parser.add_argument(
-            "--snapshot",
-            action="store_true",
-            help="Create a new Elasticsearch snapshot of the current index state which is stored in S3",
-        )
         parser.add_argument(
             "--start-datetime",
             type=datetime_command_line_argument_type(naive=False),
             help="Processes transactions updated on or after the UTC date/time provided. yyyy-mm-dd hh:mm:ss is always "
             "a safe format. Wrap in quotes if date/time contains spaces.",
+            metavar="",
         )
         parser.add_argument(
             "--skip-delete-index",
             action="store_true",
-            help="When creating a new index skip the step that deletes the old indexes and swaps the aliases. "
-            "Only used when --create-new-index is provided.",
+            help="When creating a new index, skip the steps that delete old indexes and swap aliases. "
+            "Only applicable when --create-new-index is provided.",
         )
         parser.add_argument(
             "--load-type",
             type=str,
-            help="Select which type of load to perform, current options are transactions or awards.",
+            required=True,
+            help="Select which data the ETL will process.",
             choices=["transactions", "awards"],
-            default="transactions",
         )
         parser.add_argument(
-            "--idle-wait-time",
+            "--processes",
+            type=int,
+            help="Number of parallel processes to operate. psycopg2 kicked the bucket with 100",
+            default=10,
+            choices=range(1, 71),
+            metavar="[1-70]",
+        )
+        parser.add_argument(
+            "--partitions",
             type=int,
-            help="Time in seconds the ES index process should wait before looking for a new CSV data file.",
-            default=60,
+            help="Set how much the data are partitioned. "
+            "NOTICE: large datasets may need indexes or other performance "
+            " optimizations if this is changed",
+            default=10000,
+            metavar="(default: 10,000)",
```

Comment:
> Since the data volume will be variable with each incremental run, or time some time to query beforehand, we should give them the ideal batch size as a parameter

That was the original design. Unfortunately for performance reasons I changed the parameter to number of partitions so that an index could be created for the modulo. (see help text). With the current SQL there is no way to get the extraction remotely close to functional without having an index scan on the materialized view. I do have test SQL which runs the partitioning on `transaction_normalized` (for transaction ETL) which helps even more.  I can circle back around on this after we optimize the SQL.

Reply:
Where's the index? Was there a django migration for it? Is it an index like `id % 10000`, with it specifically hard-coded to 10000?

This may be more of a case for _generating_ partition clauses dynamically. E.g. if we know 
1. we have 112,424,233 transactions to process
2. Our partition column has no nulls
3. and we know we want batch sizes of 10,000 records max

```python
def list_sql_partition_clauses(lower_bound: int, upper_bound: int, batch_size: int, partition_col: str) -> List[str]:
    step =(upper_bound - lower_bound) // batch_size
    for i in range(lower_bound, upper_bound + step, step):
        yield f"WHERE {partition_col} >= {i} AND {partition_col} < {i + step}"
```

Partitions wouldn't be exactly 10,000 items, it'd be a max. And there would very likely be skew based on the range of IDs updated for incremental loads. But there's a better chance that you'd get "fuller" batches than on smaller data volumes.

Reply:
Index was manually created when performance testing in a different branch. Since the database schema is up-in-the-air I didn't include that here. Example SQL

```sql
CREATE INDEX temp_idx_dev_5856_test_mod_10k ON temp_transaction_delta_view_table((mod(transaction_id, 10000::bigint)) int8_ops)
```

Not sure I'm following the partition clause generation. The script will already find the number of records to process. Originally I had a little equation to determine the partitions based on a desired partition size (and it would reduce the partitions a little to make them more even. e.g. with 101 records to process and a requested partition size of 10, instead of 10 partitions of ~10 and 1 of ~1, it would make 11 partitions with ~9 records each)